### ISSUE TITLE
Add-AzureRmAccount : A task was canceled.

### ISSUE THREAD

**Issue by GuofuZhong** [Microsoft/vsts-tasks#7049]
## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

## Environment
VSTS 
release definition:https://msasg.visualstudio.com/Bing_Ads/AdsUCM/_release?_a=releases&definitionId=447
release:
https://msasg.visualstudio.com/Bing_Ads/AdsUCM/_apps/hub/ms.vss-releaseManagement-web.cd-release-progress?releaseId=63514&_a=release-environment-logs&environmentId=346958
https://msasg.visualstudio.com/Bing_Ads/AdsUCM/_apps/hub/ms.vss-releaseManagement-web.cd-release-progress?releaseId=62994&_a=release-environment-logs&environmentId=345052

- Agent - Hosted or Private: 
private agent:Yes
version:2.129.1
OS: win server 2016.

## Issue Description
we are using the VSTS multiplexer to run multiple azure db deployment scripts in parallel.
Sometimes we confront the exception:Add-AzureRmAccount : A task was canceled.

### Error logs

`Add-AzureRmAccount : A task was canceled.`
`At E:\VSTSAGENTVM022-008\_work\r2\a\Common\LoginAzureRM.ps1:60 char:5`
`+     &"Login-AzureRmAccount" @parameters |Out-Null`
`+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`
 `   + CategoryInfo          : CloseError: (:) [Add-AzureRmAccount], TaskCanceledException`
`    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Profile.AddAzureRMAccountCommand`

**Comment by yiqzhou** [comment]
1. I agree that this is something about with Azure RM login. Is there any limitation of setting up multiple connections to Azure from the same IP? Have you seen similar cases? I got an error messages below:
2018-04-23T06:15:12.9970535Z ##[error]Get-AzureRmSqlServerFirewallRule : Your Azure credentials have not been set up or have expired, please run Login-AzureRMAccount to set up your Azure credentials. https://msasg.visualstudio.com/DefaultCollection/Bing_Ads/_release?releaseId=63523&_a=release-logs Phase: Deploy Databases, Multiplier: UcmKeywordOverlapping, Step: Remove Firewall Rule

2. I opened system.debug and found that in a multiple configuration agent phase, when one thread of the phase gets '[error]Add-AzureRmAccount : A task was canceled', other running threads logged '##[debug]Re-evaluate condition on job cancellation for step: xxx.' and got cancelled.
I thought failure in one thread will not stop process of other threads. Is there anything different in this case?

**Comment by GuofuZhong** [comment]
@KrishnaAdityaB  We still get this error from multiple build or release pipeline.
How to avoid this problem?

**Comment by rokmicrosoft** [comment]
This is on Azure powershell team and they are working on it https://github.com/Azure/azure-powershell/issues/6165

**Comment by niadak** [comment]
Corresponding Azure PowerShell issue is fixed. 
@GuofuZhong Can you install latest Azure PS and check if the issue is resolved.

**Comment by niadak** [comment]
@GuofuZhong , I am closing the issue as the root cause of the problem is fixed in Azure PowerShell side. Installing latest Azure PS should resolve the issue. Let us know if you need any help on this issue.

**Comment by PremRaju** [comment]
I am provisioning all resources using powershell through pipeline frequently i am getting above issue and some time it will pass.
as i see solution above Which Version of Powershell will overcome this issue and upgrade will it impact any other resource provision ??
how to upgrade the version uninstall and install the latest version? or Update-Module AzureRM?

**Comment by kmkumaran** [comment]
@PremRaju  - There was an issue in ARM which manifested as TaskCancelledException. You probably might have hit the same. The issue is now mitigated. You can find the details at https://blogs.msdn.microsoft.com/vsoservice/?p=17145

**Comment by PremRaju** [comment]
@Team,
i need your input whenever we found root cause for this issue.

Thanks lot for your input.
Prem.

**Comment by samuelbcollie** [comment]
I'm having a similar issue with "Get-AzureRMWebAppSlot" and "Switch-AzureRMWebAppSlot". I'm using a hosted agent, but I was able to upgrade AzureRM thanks to this [stack overflow post](https://stackoverflow.com/questions/50375331/upgrade-azurerm-powershell-on-hosted-2017-agent-vsts-visual-studio-team-servi). The only problem is powershell 6.3.0 and above introduce another issue #7601. If I could determine why I'm having timeouts, then I wouldn't need to upgrade. Is the work around suggested for 7601 my only option right now?

**Comment by SumiranAgg** [comment]
@samuelbcollie You can build a retry logic around your command-lets if you dont want to upgrade powershell version. Otherwise, if you upgrade powershell version, the workaround could be adding below commands before your script:
Clear-AzureRmContext -Scope Process
Clear-AzureRmContext -Scope CurrentUser
Disable-AzureRmContextAutosave
In the meanwhile you can follow this ticket https://github.com/Azure/azure-powershell/issues/6680 on azure powershell team regarding this issue.

**Comment by PremRaju** [comment]
Hello Team,
still i am getting issues on Get-AzureRmWebAppPublishingProfile  Login-AzureRmAccount : Add-AzureRmAccount : A task was canceled.
its randomly appearing one or other reasons.not frequently..

** I don't find Permanent Solution anywhere.**

Logs are like below..
powershell.exe : Get-AzureRmWebAppPublishingProfile : A task was canceled.
At C:\Jenkinsworkspace\RefDataService@tmp\durable-74699963\powershellWrapper.ps1:3 char:1
+ & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -File ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Get-AzureRmWebA...k was canceled.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
At C:\Jenkinsworkspace\RefDataService\src\DevOps\DeploymentScripts\msdeployfunctionapp.ps1:11 char:8
+ $xml = Get-AzureRmWebAppPublishingProfile -Name $AppName `
+        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    + CategoryInfo          : CloseError: (:) [Get-AzureRmWebAppPublishingProfile], TaskCanceledException

    + FullyQualifiedErrorId : Microsoft.Azure.Commands.WebApps.Cmdlets.WebApps.GetAzureWebAppPublishingProfileCmdlet
===================================================================

powershell.exe : Login-AzureRmAccount : A task was canceled.
At C:\Jenkinsworkspace\ValidationService@tmp\durable-b0661ed8\powershellWrapper.ps1:3 char:1
+ & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -File ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Login-AzureRmAc...k was canceled.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

**Comment by SumiranAgg** [comment]
@PremRaju If it is a powershell script, you can upgrade to latest powershell version.
In the meanwhile, as per comment by @kmkumaran this issue might be on ARM end. To debug can you share your VS account region, subscription ID and client IP(if possible).

**Comment by samuelbcollie** [comment]
After testing the release two more times, I'm getting the log in error again. It occurred in a different task both times.

**Comment by SumiranAgg** [comment]
@PremRaju You can share the requested details at RM_Customer_Queries@microsoft.com
@samuelbcollie Can you share debug logs for the failure and your VS account region, agent region, subscription ID and client IP(if possible) here or at RM_Customer_Queries@microsoft.com

**Comment by akhilvatts** [comment]
Facing the same issue at our end too.
+1

**Comment by SumiranAgg** [comment]
Closing this issue since no activity. Please feel free to reopen if required.
